src/LICM.cpp

```diff
@@ -0,0 +1,150 @@
+#include "LICM.h"
+#include "ExprUsesVar.h"
+#include "IREquality.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Scope.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::map;
+
+// Is it safe to lift an Expr out of a loop (and potentially across a device boundary)
+class CanLift : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const Call *op) {
+        if (!op->is_pure()) {
+            result = false;
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+
+    void visit(const Load *op) {
+        result = false;
+    }
+
+    void visit(const Variable *op) {
+        if (varying.contains(op->name)) {
+            result = false;
+        }
+    }
+
+    const Scope<int> &varying;
+
+public:
+    bool result {true};
+
+    CanLift(const Scope<int> &v) : varying(v) {}
+};
+
+// Lift pure loop invariants to the top level. Applied independently
+// to each loop.
+class LiftLoopInvariants : public IRMutator {
+    using IRMutator::visit;
+
+    Scope<int> varying;
+
+    bool can_lift(const Expr &e) {
+        CanLift check(varying);
+        e.accept(&check);
+        return check.result;
+    }
+
+    bool should_lift(const Expr &e) {
+        if (!can_lift(e)) return false;
+        if (e.as<Variable>()) return false;
+        if (e.as<Broadcast>()) return false;
+        if (is_const(e)) return false;
+        return true;
+    }
+
+    void visit(const Let *op) {
+        varying.push(op->name, 0);
+        IRMutator::visit(op);
+        varying.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) {
+        varying.push(op->name, 0);
+        IRMutator::visit(op);
+        varying.pop(op->name);
+    }
+
+    void visit(const For *op) {
+        varying.push(op->name, 0);
+        IRMutator::visit(op);
+        varying.pop(op->name);
+    }
+
+public:
+
+    using IRMutator::mutate;
+
+    Expr mutate(const Expr &e) {
+        if (should_lift(e)) {
+            auto it = lifted.find(e);
+            if (it == lifted.end()) {
+                string name = unique_name('t');
+                lifted[e] = name;
+                return Variable::make(e.type(), name);
+            } else {
+                return Variable::make(e.type(), it->second);
+            }
+        } else {
+            return IRMutator::mutate(e);
+        }
+    }
+
+    map<Expr, string, IRDeepCompare> lifted;
+};
+
+class LICM : public IRMutator {
+    using IRVisitor::visit;
+
+    bool in_gpu_loop {false};
+
+    void visit(const For *op) {
+        bool old_in_gpu_loop = in_gpu_loop;
```

Comment:
Y'know, we use this pattern pervasively enough in mutators that we can probably justify a templated helper for save_old+set_new+restore_old...

Reply:
I haven't been able to come up with anything that actually reads as clearer, e.g:

```
class Foo : public IRMutator {
    MutatorState<bool> in_gpu_loop;

    void visit(const For *op) {
        auto old_in_gpu_loop = in_gpu_loop.push(op->for_type == ...);

        if (in_gpu_loop && !old_in_gpu_loop) {
            ...
        }
    
...

```

though automatically restoring the old value would be nice from a bug-avoidance perspective

Reply:
```
class Foo : public IRMutator {
    void visit(const For *op) {
        SaveState<bool> in_gpu_loop(op->for_type == ...);
        if (in_gpu_loop && !in_gpu_loop.previous()) {
            ...
        }
    }
```

Reply:
Doesn't there need to be a class member to actually hold the state?

Reply:
Yep. Should have proofread before posting.